soa-derive-internal/src/slice.rs

```diff
@@ -153,22 +153,47 @@ pub fn derive(input: &Input) -> TokenStream {
             /// Similar to [`
             #[doc = #slice_name_str]
             /// ::get()`](https://doc.rust-lang.org/std/primitive.slice.html#method.get).
-            pub fn get(&self, i: usize) -> Option<#ref_name> {
-                if self.is_empty() || i >= self.len() {
-                    None
-                } else {
-                    Some(#ref_name {
-                        #(#fields_names_1: self.#fields_names_2.get(i).unwrap(),)*
-                    })
-                }
+            pub fn get<'b, I>(&'b self, index: I) -> Option<I::RefOutput>
+            where
+                I: ::soa_derive::SoAIndex<#slice_name<'b>>,
+                'a: 'b
+            {
+                let slice: #slice_name<'b> = self.reborrow();
+                index.get(slice)
             }
 
             /// Similar to [`
             #[doc = #slice_name_str]
             /// ::get_unchecked()`](https://doc.rust-lang.org/std/primitive.slice.html#method.get_unchecked).
-            pub unsafe fn get_unchecked(&self, i: usize) -> #ref_name {
-                #ref_name {
-                    #(#fields_names_1: self.#fields_names_2.get_unchecked(i),)*
+            pub unsafe fn get_unchecked<'b, I>(&'b self, index: I) -> I::RefOutput
+            where
+                I: ::soa_derive::SoAIndex<#slice_name<'b>>,
+                'a: 'b
+            {
+                let slice: #slice_name<'b> = self.reborrow();
+                index.get_unchecked(slice)
+            }
+
+            /// Similar to [`std::ops::Index` trait](https://doc.rust-lang.org/std/ops/trait.Index.html) on
+            #[doc = #slice_name_str]
+            /// .
+            /// This is required because we cannot implement that trait.
+            pub fn index<'b, I>(&'b self, index: I) -> I::RefOutput
+            where
+                I: ::soa_derive::SoAIndex<#slice_name<'b>>,
+                'a: 'b
+            {
+                let slice: #slice_name<'b> = self.reborrow();
+                index.index(slice)
+            }
+
+            /// Reborrows the slices in a more narrower lifetime
```

Comment:
```suggestion
            /// Reborrows the slices in a more narrow lifetime
```

Reply:
In this context `a narrower lifetime` sound better to me than `a more narrow lifetime` (and yes, `more narrower` is not correct). What do you think?

Reply:
sounds good!